### PR TITLE
Resource Loader

### PR DESCRIPTION
**Pull request by Alan5142**
* PoC resource loader

### CODE REVIEW

File: src/engine_core/base/src/Common.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    \date 2024-10-13
+    \brief Common definitions
+*/
+
+#pragma once
+
+#include "Platform.hpp"
+
+// By default export the libs.
+#ifndef HUSH_EXPORT_LIB
```
Comment (Alan5142) [inline src/engine_core/base/src/Common.hpp:12]: Woops, I left this here.
I plan to add support for loadable modules in the future. 
This is not relevant for this PR.

File: src/engine_core/filesystem/src/FileSystem.hpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        IFileSystem &operator=(const IFileSystem &) = delete;
+        IFileSystem &operator=(IFileSystem &&) = default;
+
+        virtual ~IFileSystem() = default;
+
+        /// Opens a file from a specific path.
+        /// @param path Path of the file to open
+        /// @param mode Mode to open the file
+        /// @return A result with a pointer to the file.
+        virtual Result<std::unique_ptr<IFile>, IFile::EError> OpenFile(std::filesystem::path vfs_path,
```
Comment (Kyn21kx) [inline src/engine_core/filesystem/src/FileSystem.hpp]: Nitpick:
Wrong parameter naming convention `vfs_path` should be `vfsPath`

File: src/engine_core/filesystem/src/filesystem/CFileSystem/CFileSystem.cpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    else if (mode == EFileOpenMode::Write)
+    {
+        modeStr = {'w', 'b', '\0'};
+    }
+    else if (mode == EFileOpenMode::ReadWrite)
+    {
+        modeStr = {'r', '+', 'b', '\0'};
+    }
+
+    const auto real_path = mRoot / path;
```
Comment (Kyn21kx) [inline src/engine_core/filesystem/src/filesystem/CFileSystem/CFileSystem.cpp]: I would highly advice against the use of `auto` outside of C++ casts (i.e `auto* foo = static_cast<Foo*>(bar);`) where the type can clearly be inferred at first glance

File: src/engine_core/filesystem/src/filesystem/CFileSystem/CFileSystem.cpp [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+    FILE *file = nullptr;
+
+    const auto real_path_str = real_path.generic_string();
+
+    if (auto error = fopen_s(&file, real_path_str.c_str(), modeStr.data()); error != 0)
```
Comment (Kyn21kx) [inline src/engine_core/filesystem/src/filesystem/CFileSystem/CFileSystem.cpp]: Same, `auto` use

**Review by Kyn21kx — APPROVED** [review]
Just a few minor details with coding conventions

File: src/engine_core/filesystem/src/FileSystem.hpp (hunk 2 above)
Comment (Alan5142) [reply src/engine_core/filesystem/src/FileSystem.hpp]: Changed

File: src/engine_core/filesystem/src/filesystem/CFileSystem/CFileSystem.cpp (hunk 3 above)
Comment (Alan5142) [reply src/engine_core/filesystem/src/filesystem/CFileSystem/CFileSystem.cpp]: I changed this one to std::filesystem::path.
For other usages, such as iterators, I suggest using auto as the type can be long (and the context suggest is an iterator)

File: src/engine_core/filesystem/src/filesystem/CFileSystem/CFileSystem.cpp (hunk 4 above)
Comment (Alan5142) [reply src/engine_core/filesystem/src/filesystem/CFileSystem/CFileSystem.cpp]: Replaced with errno_t

**Pull request merged**